add applied fairness analysis with yardstick (#48)
tidymodels/tidymodels.org: learn/work/fairness-detectors/index.qmd

---
title: "Are GPT detectors fair? A machine learning fairness case study"
categories:
  - yardstick
  - metrics
type: learn-subsection
weight: 6
description: |

>   Releases of language models are often soon followed up with release of detectors, claiming to detect whether inputted text was written by a a human or a given language model. A 2022 study argues that GPT detectors disproportionately classify real writing from non-native English writers as AI-generated.

Reviewer: ```suggestion
  Releases of language models are often soon followed up with detectors, claiming to detect whether text was written by a human or a given language model. A 2022 study argues that GPT detectors disproportionately classify real writing from non-native English writers as AI-generated.
```

toc: true
toc-depth: 2
include-after-body: ../../../resources.html
---

------------------------------------------------------------------------

```{r}
#| label: "setup"
#| include: false
#| message: false
#| warning: false
source(here::here("common.R"))
```

```{r}
#| label: "load"
#| include: false
library(tidymodels)
pkgs <- c("tidymodels")

theme_set(theme_bw() + theme(legend.position = "top"))
```

> The popularity and capability of language models like GPT-4, which generate human-like text based on simple prompts, has rapidly increased in recent years. The releases of more powerful next-generation models are often accompanied by the development of GPT *detectors*, which are machine learning models intended to estimate the probability that inputted text was generated by a GPT.

Reviewer: ```suggestion
The popularity and capability of large language models like GPT-4, which generate human-like text based on simple prompts, has rapidly increased in recent years. The releases of more powerful next-generation models are often accompanied by the development of GPT *detectors*, which are machine learning models intended to estimate the probability that inputted text was generated by a GPT.
```

In 2022, a group of researchers collected a corpus of essays, some of which were written by real humans and others that were generated using various language models. Notably, some of the essays from real humans were written by people who did not write in English natively, while others were written by people who did.

::: callout-note
What does it mean to be a "native writer" of a language? Keep an eye out for this idea later on.
:::

> The researchers passed each essay to a number of marketed GPT detectors, recording the predicted probabilities that each essay was written by a human or GPT model. Juxtaposing detector predictions for papers written by native and non-native English writers, Liang et al. argued in their paper [*GPT detectors are biased against non-native English writers*](https://doi.org/10.1016/j.patter.2023.100779) that GPT detectors disproportionately classify real writing from non-native English writers as AI-generated. Some news media channels that cite the study use it as evidence that GPT detector models behave unfairly. The statements 1) *GPT detectors are biased against non-native English writers*, 2) *GPT detectors disproportionately classify real writing from non-native English writers as AI-generated*, and 3) *GPT detector models behave unfairly* are only equivalent, though, given a shared set of social values surrounding what *fairness* means.

Reviewer: ```suggestion
The researchers passed each essay to a number of marketed GPT detectors, recording the predicted probabilities that each essay was written by a human or GPT model. Juxtaposing detector predictions for papers written by native and non-native English writers, Liang et al. argued in their paper [*GPT detectors are biased against non-native English writers*](https://doi.org/10.1016/j.patter.2023.100779) that GPT detectors disproportionately classify real writing from non-native English writers as AI-generated. Some news media channels that cite the study use it as evidence that GPT detector models behave unfairly. The statements 

1. GPT detectors are biased against non-native English writers, 
2. GPT detectors disproportionately classify real writing from non-native English writers as AI-generated, and 
3. GPT detector models behave unfairly

are only equivalent, though, given a shared set of social values surrounding what *fairness* means.
```

> Perhaps the core question that machine learning fairness as a research field has tried to address is exactly what a machine learning model acting fairly entails. As a recent primer notes, "\[t\]he rapid growth of this new field has led to wildly inconsistent motivations, terminology, and notation, presenting a serious challenge for cataloging and comparing definitions" (Mitchell et al. 2021). Fairness is a social construct, and means different things to different people; the operationalization of fairness as a quantitative metric is not at all trivial. The aim of evaluating a machine learning system's fairness is thus to state clearly one's definition of fairness, critique what that definition means in the problem context, and assess fairness under the chosen definition rigorously. The tidymodels packages in R provide a suite of functionality to support this kind of analysis, and we'll use the data from Liang et al. as an example to demonstrate a fairness-oriented analysis of model performance.

Reviewer: Is it possible to link to the papers at the bottom of the page using citations? Or not worth the trouble?
https://quarto.org/docs/authoring/footnotes-and-citations.html

Author: Thank you for the nudge—this was the right move. Done in [4bbd9f9](https://github.com/tidymodels/tidymodels.org/pull/48/commits/4bbd9f9f19be00d627ab6cfad3d2b2ef558b75e9). :)

Adding `native` into the tabulation:

```{r count-native-kind-pred}
detectors %>% 
  count(native, kind, .pred_class)
```

Note that, for essays written by AI (`kind == "AI"`), the `native` variable isn't well-defined. Those entries thus match the numbers from the above table.

For essays written by `Human`s, a plot perhaps better demonstrates the disparity:

```{r plot-human-preds}
detectors %>% 
  filter(!is.na(native)) %>%
  ggplot(aes(x = native, fill = .pred_class)) +
  geom_bar() +
  labs(x = "Native English Writer", fill = "Predicted Class")
```

Most of the essays written by non-native English writers are incorrectly classified as written by AI, while nearly all of the essays written by native English writers are correctly classified as written by humans. The same effect can be seen in the underlying probability distributions:

```{r plot-human-preds-by-native}

> detectors %>%

Reviewer: ```suggestion
#| fig-height: 9
detectors %>%
```

Trying to make this one a little taller

  filter(!is.na(native)) %>%
  mutate(native = case_when(native == "Yes" ~ "Native English writer",
                            native == "No" ~ "Non-native English writer")) %>%
  ggplot(aes(.pred_AI, fill = native)) +
  geom_histogram(bins = 30, show.legend = FALSE) +
  facet_wrap(vars(native), scales = "free_y", nrow = 2) +
  labs(x = "Predicted Probability That Essay Was Written by AI")
```

> Again, note that *all of the plotted essays were written by humans*. An effective detector would thus predict a probability near to zero for all of these observations. In this plot, we see that the evidence in our initial table showing these detectors weren't performing well didn't tell the whole story. They perform *quite* well for native English writers, actually. For non-native English writers, though, they perform terribly.

Reviewer: ```suggestion
Again, note that *all of the plotted essays were written by humans*. An effective detector would thus predict a probability near zero for all of these observations. In this plot, we see that the evidence in our initial table showing these detectors weren't performing well didn't tell the whole story. They perform *quite* well for native English writers, actually. For non-native English writers, though, they perform terribly.
```

In this example, a GPT detector satisfies equal opportunity when the detector correctly classifies human-written text as human-written and incorrectly classifies human-written text as generated by AI at the same rate for both native English writers and non-native English writers.

::: callout-note
This definition does not consider the predictions based on essays generated by AI.
:::

To calculate equal opportunity with yardstick, we use the `equal_opportunity()` function:

```{r ep-opp-native}
equal_opportunity_by_native <- equal_opportunity(by = native)
```

<!-- TODO: print out the following function once the metric print method is merged into yardstick -->

The function `equal_opportunity_by_native()` is a yardstick metric function like any other, except it knows to temporarily group by and summarize across a data-column called `native`. Applying it:

```{r eq-opp}
detectors %>%

>   filter(kind == "Human") %>%

Author: We shouldn't have to include this step, I think, as this metric doesn't use the predictions on essays generated by AI. For some reason, this errors out at the moment.

Author: Closing in light of looking into this at https://github.com/tidymodels/yardstick/issues/464.

  group_by(detector) %>%
  equal_opportunity_by_native(truth = kind, estimate = .pred_class, event_level = "second") %>%
  arrange(.estimate) %>%
  head(3)
```

The detectors with estimates closest to zero are most fair, by this definition of fairness.

::: callout-important

> Why do we need to specify `event_level = "second"`?

Reviewer: I suggest you add a link here so folks can look this up. What's the best link that explains this in the yardstick documentation?

Author: This is a good idea. It doesn't appear to me that there are any dedicated help-files in yardstick docs about this, so maybe we link to the "Relevant level" section of the yardstick function behind the scenes in `equal_opportunity()`? At https://yardstick.tidymodels.org/dev/reference/sens.html#relevant-level

Reviewer 2: yes, that appears to be the best place to link to at the moment

:::

Given this set of moral values, our analysis would offer a different set of recommendations for which detector to use.

## Balancing two notions of fairness

> Instead, imagine that we both feel that it is unfair for GPT-generated work to go undetected *and* believe that it's unfair for a GPT detector to disproportionately classify human-written work from non-native English writers as GPT-generated.

Reviewer: ```suggestion
Instead of either of the options we have considered so far, imagine that we both feel that it is unfair for GPT-generated work to go undetected *and* believe that it's unfair for a GPT detector to disproportionately classify human-written work from non-native English writers as GPT-generated.
```

  arrange(desc(.estimate)) %>%
  head(3)

detectors %>%
  filter(kind == "Human", detector %in% performant_detectors$detector) %>%
  group_by(detector) %>%
  equal_opportunity_by_native(truth = kind, estimate = .pred_class, event_level = "second") %>%
  arrange(.estimate)
```

We could also reverse the process, reflecting the belief that it is more unfair to disproportionately classify human-written work from non-native English writers as GPT-generated than it is to pass on output from GPTs as one's own work. We first set a threshold based on `equal_opportunity()`, then choose the most performant model by `roc_auc()`:

```{r eq-opp-dets}
equal_opportunity_detectors <- 
  detectors %>%
  filter(kind == "Human") %>%
  group_by(detector) %>%
  equal_opportunity_by_native(truth = kind, estimate = .pred_class, event_level = "second") %>%
  arrange(.estimate) %>%
  head(3)

detectors %>%
  filter(detector %in% equal_opportunity_detectors$detector) %>%
  group_by(detector) %>%
  roc_auc(truth = kind, .pred_AI) %>%
  arrange(desc(.estimate))
```

In this example, changing the prioritization of the criteria results in a different set of recommended models.

# Choosing a detector

> In the preceding section, we saw that the recommended detector model resulting from our analysis depended on our moral values; the mathematical notion of fairness appropriate for a given analysis follows from the problem context. In this way, no statistical model is objectively more fair than another; our assessment of fairness depends on our personally held ideas of fairness.

Reviewer: ```suggestion
In the preceding section, we saw that the recommended detector we identify depends on our moral values; the mathematical notion of fairness appropriate for a given analysis follows from the problem context. In this way, no statistical model is objectively more fair than another; our assessment of fairness depends on our personally held ideas of fairness.
```

```{r rocs}
#| include: false
rocs <- 
  detectors %>%
  group_by(detector) %>%
  roc_auc(truth = kind, .pred_AI) %>%
  arrange(desc(.estimate))
```

As for the problem of GPT detection, while each stakeholder might find that some models are more fair than others, even the most fair models recommended in each approach are quite unfair. For instance, from the first stakeholder's perspective, even though it's the most performant model available, `r rocs$detector[1]`'s `roc_auc()` of `r round(rocs$.estimate[1], 3)` leaves much to be desired; a stakeholder ought to consider the potential harms resulting from the substantial number of errors made by this model when applied in context.

> This analysis only considered one fairness metric, `equal_opportunity()`. We could have attempted to apply either of the other two fairness metrics included in yardstick 1.3.0, `equalized_odds()` and `demographic_parity()`, or a custom fairness metric. Are those two other metrics well-defined for this problem? The interests of which stakeholders are best represented by those metrics? Would they result in yet another set of discordant recommendations?

Reviewer: ```suggestion
This analysis only considered one fairness metric, `equal_opportunity()`. We could have attempted to apply either of the other two fairness metrics included in yardstick 1.3.0, `equalized_odds()` and `demographic_parity()`, or a custom fairness metric. Are those two other metrics well-defined for this problem? Which stakeholders' interests are best represented by those metrics? Would they result in yet another set of discordant recommendations?
```

> We also did not consider how the outputs of a chosen model be used. If a student's work is classified as written by a GPT, what happens then? Would a misclassification be more harmful for one type of student than another? Could an instructor trust model output more readily for one type of student than another? Answers to these questions are a necessary component of a complete fairness analysis and, just like the choice of metric, depend heavily on the problem context.

Reviewer: ```suggestion
We also did not consider how the outputs of a chosen model be used. If a student's work is classified as written by a GPT model, what happens then? Would a misclassification be more harmful for one type of student than another? Could an instructor trust model output more readily for one type of student than another? Answers to these questions are a necessary component of a complete fairness analysis and, just like the choice of metric, depend heavily on the problem context.
```

---
title: "Are GPT detectors fair? A machine learning fairness case study"
categories:
  - yardstick

>   - metrics

Reviewer 2: is `-metrics` needed here? Right now it is the only one being used.

I would be happy to change it to be fairness related such that once we have more fairness stuff it is nicely labeled

Author: Good point. Will do this for the other, too.

type: learn-subsection
weight: 6
description: |
  Releases of language models are often soon followed up with detectors, claiming to detect whether text was written by a human or a given language model. A 2023 study argues that GPT detectors disproportionately classify real writing from non-native English writers as AI-generated.
toc: true
toc-depth: 2
include-after-body: ../../../resources.html
bibliography: refs.bib
---

------------------------------------------------------------------------

```{r}
#| label: "setup"
#| include: false
#| message: false
#| warning: false
source(here::here("common.R"))
```

```{r}
#| label: "load"
#| include: false
library(tidymodels)
pkgs <- c("tidymodels")

theme_set(theme_bw() + theme(legend.position = "top"))
```

> The popularity and capability of large language models like GPT-4, which generate human-like text based on simple prompts, has rapidly increased in recent years. The releases of more powerful next-generation models are often accompanied by the development of GPT *detectors*, which are machine learning models intended to estimate the probability that inputted text was generated by a GPT.

Reviewer 2: The acronym "GPT" is used a lot in this article. I think it would be helpful to spell it out, AND to add a link to `GPT-4` as it is a specific service

3. GPT detector models behave unfairly

are only equivalent, though, given a shared set of social values surrounding what *fairness* means.

Perhaps the core question that machine learning fairness as a research field has tried to address is exactly what a machine learning model acting fairly entails. As a recent primer notes, "\[t\]he rapid growth of this new field has led to wildly inconsistent motivations, terminology, and notation, presenting a serious challenge for cataloging and comparing definitions" [@mitchell2021]. Fairness is a social construct, and means different things to different people; the operationalization of fairness as a quantitative metric is not at all trivial. The aim of evaluating a machine learning system's fairness is thus to state clearly one's definition of fairness, critique what that definition means in the problem context, and assess fairness under the chosen definition rigorously. The tidymodels packages in R provide a suite of functionality to support this kind of analysis, and we'll use the data from Liang et al. as an example to demonstrate a fairness-oriented analysis of model performance.

> After loading needed packages, we'll conduct an exploratory analysis of the study's data. Once we have a sense for the distributions of relevant variables, we'll explore the different ways that we could try to measure fairness in this problem context. We will see that different ways of measuring fairness encode social values held by different stakeholders.

Reviewer 2: ```suggestion
After loading the needed packages, we'll conduct an exploratory analysis of the study's data. Once we have a sense for the distributions of relevant variables, we'll explore the different ways that we could try to measure fairness in this problem context. We will see that different ways of measuring fairness encode social values held by different stakeholders.
```

> After loading needed packages, we'll conduct an exploratory analysis of the study's data. Once we have a sense for the distributions of relevant variables, we'll explore the different ways that we could try to measure fairness in this problem context. We will see that different ways of measuring fairness encode social values held by different stakeholders.

Reviewer 2: ```suggestion
After loading needed packages, we'll conduct an exploratory analysis of the study's data. Once we have a sense of the distributions of relevant variables, we'll explore the different ways that we could try to measure fairness in this problem context. We will see that different ways of measuring fairness encode social values held by different stakeholders.
```

Reviewer: FWIW I think "for" would read better here

```{r plot-human-preds-by-native-detector}
detectors %>%
  filter(!is.na(native)) %>%
  mutate(native = case_when(native == "Yes" ~ "Native English writer",
                            native == "No" ~ "Non-native English writer")) %>%
  ggplot(aes(.pred_AI, fill = native)) +
  geom_histogram(bins = 30) +
  facet_grid(vars(native), vars(detector), scales = "free_y") +
  labs(x = "Predicted Probability That Essay Was Written by AI") +
  scale_x_continuous(breaks = c(0, 1)) +
  theme(legend.position = "none")
```

Each column in this plot reflects roughly the same story as the plot that aggregates across detectors; the detectors work *very* well at correctly classifying real writing from native English writers, yet classify writing from non-native English writers incorrectly at least as often as they do so correctly.

::: callout-note
Explore the source data further. Where are the essays written by native English writers collected from? How about non-native? What does our usage of "native writer" mean in this context, then?

> :::

Reviewer 2: This is more of a question to the user rather than a note. We can use a [title inside the callout block](https://quarto.org/docs/authoring/callouts.html) to highlight that

```suggestion
::: callout-note
## Question for reflection
Explore the source data further. Where are the essays written by native English writers collected from? How about non-native? What does our usage of "native writer" mean in this context, then?
:::
```

Author: Woahhh! TIL!

# Fairness assessment

> The 1.3.0 release of yardstick introduced a suite of functionality for fairness-oriented analysis of machine learning models. The release includes a tool to make fairness metrics, `new_groupwise_metric()`, as well as three outputs of that tool giving canonical fairness metrics: `equal_opportunity()`, `equalized_odds()`, and `demographic_parity()`. To better understand how to use metrics to quantify model fairness, we'll consider what a fairness assessment of this model would look like from the perspective of three different stakeholders.

Reviewer 2: We should link to the release when it appears

```suggestion
The [1.3.0 release of yardstick](https://yardstick.tidymodels.org/news/index.html#yardstick-130) introduced a suite of functionality for fairness-oriented analysis of machine learning models. The release includes a tool to make fairness metrics, `new_groupwise_metric()`, as well as three outputs of that tool giving canonical fairness metrics: `equal_opportunity()`, `equalized_odds()`, and `demographic_parity()`. To better understand how to use metrics to quantify model fairness, we'll consider what a fairness assessment of this model would look like from the perspective of three different stakeholders.
```

## Effective detection, group-blind

> Imagine, first, the position that the most fair detection model is one that reliably differentiates between essays written by humans or generated by GPTs, regardless of problem context. From this perspective, the most fair model is the model that detects GPT-generated essays most effectively; it is unfair to pass on an essay written by a GPT as one's own work. When analyzing this data, a stakeholder with this perspective would ignore the `native` variable in their analysis.

Reviewer 2: ```suggestion
Imagine, first, the position that the most fair detection model is one that reliably differentiates between essays written by humans or generated by GPTs, regardless of the problem context. From this perspective, the most fair model is the model that detects GPT-generated essays most effectively; it is unfair to pass on an essay written by a GPT as one's own work. When analyzing this data, a stakeholder with this perspective would ignore the `native` variable in their analysis.
```

:::

<!-- TODO: remove the /dev from the above link once this yardstick vignette makes it to CRAN -->

From this perspective, the models with the highest `roc_auc()` estimates are the most fair.

The fairness assessment for this stakeholder doesn't need any of the functionality newly introduced in yardstick 1.3.0. An analysis that reconciles the role of the `native` variable in these models' predictions will, though.

## Fair prediction on human-written essays

Now, consider that our only priority was to correctly classify human-written text as human-written and incorrectly classify human-written text as generated by AI at the same rate for both native English writers and non-native English writers, ignoring predictions from essays generated by AI.

-   Another **student** whose work is evaluated by a detector model may take on such a perspective. This student could be a native English writer who does not want other students to be subjected to undue harm or a non-native English writer concerned with their own writing being incorrectly classified as GPT-generated.

> -   Another **instructor** of a course may feel it is unfair to disproportionately classify writing from students that are non-native English writers as GPT-generated, regardless of how effectively the model detects GPT-generated essays.

Reviewer 2: ```suggestion
-   Another **instructor** of a course may feel it is unfair to disproportionately classify writing from students who are non-native English writers as GPT-generated, regardless of how effectively the model detects GPT-generated essays.
```

The `equal_opportunity()` metric enables us to quantify the extent of this interpretation of unfairness. Equal opportunity is satisfied when a model's predictions have the same true positive and false negative rates across protected groups; a model predicts more fairly if it's equally likely to predict a positive outcome for each group.

In this example, a GPT detector satisfies equal opportunity when the detector correctly classifies human-written text as human-written and incorrectly classifies human-written text as generated by AI at the same rate for both native English writers and non-native English writers.

::: callout-note

> This definition does not consider the predictions based on essays generated by AI.

Reviewer 2: Maybe mention in this callout the metrics Sensitivity and Specificity? At first glance it might look wrong to only look at the non-AI, but such actions are not uncommon, as seen with Sensitivity and Specificity

:::

> To calculate equal opportunity with yardstick, we use the `equal_opportunity()` function:

Reviewer 2: I think it would be nice if we say that `equal_opportunity()` creates a function here, as function factories are a little less common

```suggestion
To calculate equal opportunity with yardstick, we use `equal_opportunity()` to create a metric function:
```

```{r ep-opp-native}
equal_opportunity_by_native <- equal_opportunity(by = native)
```

<!-- TODO: print out the following function once the metric print method is merged into yardstick -->

The function `equal_opportunity_by_native()` is a yardstick metric function like any other, except it knows to temporarily group by and summarize across a data-column called `native`. Applying it:

```{r eq-opp}
detectors %>%
  filter(kind == "Human") %>%
  group_by(detector) %>%

>   equal_opportunity_by_native(truth = kind, estimate = .pred_class, event_level = "second") %>%

Reviewer 2: this line, and a couple more are a little too long to fit for the website without needing scrolling
